components/org.wso2.carbon.identity.oauth.uma.grant/src/main/java/org/wso2/carbon/identity/uma/grant/GrantType.java

```diff
@@ -0,0 +1,160 @@
+import org.wso2.carbon.identity.oauth.uma.permission.service.dao.PermissionTicketDAO;
+import org.wso2.carbon.identity.oauth.uma.permission.service.model.Resource;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.ResponseHeader;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.uma.grant.connector.PolicyEvaluator;
+import org.wso2.carbon.identity.uma.grant.internal.AuthorizeServiceComponent;
+
+import java.security.Key;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Grant type for User Managed Access 2.0.
+ */
+public class GrantType extends AbstractAuthorizationGrantHandler {
+
+    private static Log log = LogFactory.getLog(GrantType.class);
+    private static Map<Integer, Key> privateKeys = new ConcurrentHashMap<>();
+
+    boolean authStatus = false;
+    public String subject;
+    String grantType = null;
+    String permissionTicket = null;
+    String idToken = null;
+    boolean isMatched = false;
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+
+        if (log.isDebugEnabled()) {
+            log.debug("In GrantType validateGrant method.");
+        }
+
+        // extract request parameters
+        RequestParameter[] parameters = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+
+        // find out grant type
+        for (RequestParameter parameter : parameters) {
+            if (UMAGrantConstants.UMA_GRANT_PARAM.equals(parameter.getKey())) {
+                if (parameter.getValue() != null) {
+                    grantType = parameter.getValue()[0];
+                }
+            }
+
+            // find out permission ticket
+            if (UMAGrantConstants.PERMISSION_TICKET.equals(parameter.getKey())) {
+                if (parameter.getValue() != null) {
+                    permissionTicket = parameter.getValue()[0];
+                }
+            }
+
+            //find out Username from the IDToken by calling oauth2utills.getID token
+            if (UMAGrantConstants.ID_TOKEN.equals(parameter.getKey())) {
+                if (parameter.getValue() != null) {
+                    idToken = parameter.getValue()[0];
+                    isMatched = true;
+                }
+            }
+        }
+
+        if (grantType != null) {
+
+            //validate grant type and permission ticket and extract subject from idToken
+            authStatus = isValidGrantType(grantType, permissionTicket, subject);
+
+            if (authStatus) {
+
+                AuthenticatedUser authenticatedUser = new AuthenticatedUser();
+                authenticatedUser.setUserName(grantType);
```

Comment:
**grantType** parameter is set as the authenticated user's username value?
Is this correct????